fuel-derive/src/asm.rs

```diff
@@ -0,0 +1,1006 @@
+
+        let content;
+        let _bracket_token = syn::bracketed!(content in input);
+
+        while !content.is_empty() {
+            let item: InstructionArgument = content.parse()?;
+            args.push(item);
+        }
+
+        // Check argument format
+        if args.len() > 4 {
+            return Err(syn::Error::new_spanned(
+                &opcode_name,
+                format!("Too many arguments: {}", args.len()),
+            ));
+        }
+
+        for arg in args.iter().rev().skip(1) {
+            if arg.is_imm() {
+                return Err(syn::Error::new_spanned(
+                    &arg.type_,
+                    "Immediate argument only allowed as last argument",
+                ));
+            }
+        }
+
+        if args.iter().map(|a| a.typeinfo().size_bits()).sum::<usize>() > 24 {
+            return Err(syn::Error::new_spanned(
+                &opcode_name,
+                "Arguments exceed 24 bits",
+            ));
+        }
+
+        Ok(Self {
+            description,
+            opcode_number,
+            opcode_name,
+            opcode_fn_name,
+            args,
+        })
+    }
+}
+impl Instruction {
+    fn has_imm(&self) -> bool {
+        self.args.last().map(|arg| arg.is_imm()).unwrap_or(false)
+    }
+
+    #[allow(clippy::arithmetic_side_effects)] // Checked in opcode construction
+    fn reserved_bits(&self) -> usize {
+        if self.has_imm() {
+            0
+        } else {
+            24 - self.args.len() * 6
+        }
+    }
+}
+
+struct InstructionList(Vec<Instruction>);
+impl Parse for InstructionList {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let mut instructions = Vec::new();
+        while !input.is_empty() {
+            let item: Instruction = input.parse()?;
+            instructions.push(item);
+        }
+        Ok(Self(instructions))
+    }
+}
+
+/// Constructor functions and theirs shorthands
+fn make_constructors(instructions: &InstructionList) -> TokenStream {
+    instructions
+        .0
+        .iter()
+        .map(
+            |Instruction {
+                 description,
+                 opcode_name,
+                 opcode_fn_name,
+                 args,
+                 ..
+             }| {
+                let strict_arguments: TokenStream = args
+                    .iter()
+                    .map(|arg| {
+                        let name = &arg.name;
+                        let type_ = &arg.type_;
+                        quote! { #name: #type_, }
+                    })
+                    .collect();
+
+                let pack_strict_arguments: TokenStream = args
+                    .iter()
+                    .enumerate()
+                    .map(|(i, arg)| {
+                        let name = &arg.name;
+                        if arg.is_imm() {
+                            quote! {
+                                packed_integer |= (#name.to_smallest_int() as u32);
```

Comment:
Places like this would be better with a single-use helper function.

Reply:
Moved most of these to smaller functions.